zjit/src/codegen.rs

```diff
@@ -353,7 +353,7 @@ fn gen_insn(cb: &mut CodeBlock, jit: &mut JITState, asm: &mut Assembler, functio
         Insn::SetGlobal { id, val, state: _ } => return Some(gen_setglobal(asm, *id, opnd!(val))),
         Insn::GetGlobal { id, state: _ } => gen_getglobal(asm, *id),
         &Insn::GetLocal { ep_offset, level } => gen_getlocal_with_ep(asm, ep_offset, level)?,
-        Insn::SetLocal { val, ep_offset, level } => return gen_setlocal_with_ep(asm, opnd!(val), *ep_offset, *level),
+        &Insn::SetLocal { val, ep_offset, level } => return gen_setlocal_with_ep(asm, opnd!(val), function.is_a(val, types::Immediate), ep_offset, level),
```

Comment:
> function.is_a(val, types::Immediate)

I feel like this kind of type checking (not just fetching the type, but checking that it's "immediate") should be part of the codegen logic (i.e. live inside `gen_setlocal_with_ep()`). You generally wouldn't want to go back and forth between `gen_insn()` and `gen_setlocal_with_ep()` too often when you read/write the logic of codegen.

One way to do that would be to give `val_type: Type` as an argument. Another way would be to just pass around a reference to `function` (which seems fine to me) if we want to use the `Function::is_a` API.